tests/test_adjoint_jacobian.py

```diff
@@ -0,0 +1,505 @@
+
+    def test_rx_gradient(self, tol, dev):
+        """Test that the gradient of the RX gate matches the known formula."""
+        a = 0.7418
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(a, wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        # circuit jacobians
+        dev_jacobian = dev.adjoint_jacobian(tape)
+        expected_jacobian = -np.sin(a)
+        assert np.allclose(dev_jacobian, expected_jacobian, atol=tol, rtol=0)
+
+    def test_multiple_rx_gradient(self, tol):
+        """Tests that the gradient of multiple RX gates in a circuit yields the correct result."""
+        dev = qml.device("lightning.qubit", wires=3)
+        params = np.array([np.pi, np.pi / 2, np.pi / 3])
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(params[0], wires=0)
+            qml.RX(params[1], wires=1)
+            qml.RX(params[2], wires=2)
+
+            for idx in range(3):
+                qml.expval(qml.PauliZ(idx))
+
+        # circuit jacobians
+        dev_jacobian = dev.adjoint_jacobian(tape)
+        expected_jacobian = -np.diag(np.sin(params))
+        assert np.allclose(dev_jacobian, expected_jacobian, atol=tol, rtol=0)
+
+    qubit_ops = [getattr(qml, name) for name in qml.ops._qubit__ops__]
+    ops = {qml.RX, qml.RY, qml.RZ, qml.PhaseShift, qml.CRX, qml.CRY, qml.CRZ, qml.Rot}
+
+    @pytest.mark.parametrize("obs", [qml.PauliX, qml.PauliY])
+    @pytest.mark.parametrize("op", ops)
+    def test_gradients(self, op, obs, tol, dev):
+        """Tests that the gradients of circuits match between the finite difference and device
+        methods."""
+        args = np.linspace(0.2, 0.5, op.num_params)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.Hadamard(wires=0)
+            qml.RX(0.543, wires=0)
+            qml.CNOT(wires=[0, 1])
+
+            op(*args, wires=range(op.num_wires))
+
+            qml.Rot(1.3, -2.3, 0.5, wires=[0])
+            qml.RZ(-0.5, wires=0)
+            qml.RY(0.5, wires=1).inv()
+            qml.CNOT(wires=[0, 1])
+
+            qml.expval(obs(wires=0))
+            qml.expval(qml.PauliZ(wires=1))
+
+        tape.execute(dev)
+
+        tape.trainable_params = set(range(1, 1 + op.num_params))
+
+        grad_F = tape.jacobian(dev, method="numeric")
+        grad_D = dev.adjoint_jacobian(tape)
+        print(grad_D, grad_F)
+
+        assert np.allclose(grad_D, grad_F, atol=tol, rtol=0)
+
+    def test_gradient_gate_with_multiple_parameters(self, tol, dev):
+        """Tests that gates with multiple free parameters yield correct gradients."""
+        x, y, z = [0.5, 0.3, -0.7]
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(0.4, wires=[0])
+            qml.Rot(x, y, z, wires=[0])
+            qml.RY(-0.2, wires=[0])
+            qml.expval(qml.PauliZ(0))
+
+        tape.trainable_params = {1, 2, 3}
+
+        grad_D = dev.adjoint_jacobian(tape)
+        grad_F = tape.jacobian(dev, method="numeric")
```

Comment:
Note that, while it is not officially deprecated yet, there is a plan to remove `tape.jacobian` in a later release of PL!

The new canonical way of computing Jacobians will be via the [`qml.gradients`](https://pennylane.readthedocs.io/en/latest/code/qml_gradients.html#transforming-tapes) package:

```pycon
>>> gradient_tapes, fn = qml.gradients.finite_diff(tape)
>>> fn(qml.execute(gradient_tapes, device, None))
```

This API is now available in PL master, and due for the upcoming v0.18 release.

Reply:
The reason for this new API:

- It separates the tape-manipulation and classical post-processing logic from evaluation of quantum devices
- The gradient transforms _themselves_ are fully differentiable, allowing for nth order derivatives.

Reply:
Thanks Josh. @trbromley Is this something you have time to change? If not, I can do it as a separate PR.

Reply:
Oh was just leaving this here more as a comment! This is currently in PL master, so changing it now won't work if the lightning tests are run against PL stable.

Reply:
@josh146, using
```python
gradient_tapes, fn = qml.gradients.finite_diff(tape)
numeric_val = fn(qml.interfaces.batch.execute(gradient_tapes, dev, None))
```
Does this assume that the tape is already suitably expanded into device-supported ops? E.g., doing this naively and I'm getting `Gate MottonenStatePreparation not supported on device lightning.qubit`. I guess the second piece is to ensure devices are responsible for their own expansion into supported gates?

Reply:
> I guess the second piece is to ensure devices are responsible for their own expansion into supported gates?

Yep, exactly right - that is my goal for the next iteration :)